src/rapidata/service/token_manager.py

```diff
@@ -0,0 +1,137 @@
+import json
+import threading
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Optional, Callable
+
+import requests
+
+from rapidata.service.credential_manager import CredentialManager
+
+
+@dataclass
+class TokenInfo:
+    access_token: str
+    expires_at: datetime
+    token_type: str = "Bearer"
+
+    @property
+    def auth_header(self):
+        return f"{self.token_type} {self.access_token}"
+
+    @property
+    def time_until_refresh(self) -> float:
+        return (self.expires_at - datetime.now()).total_seconds()
+
+    @property
+    def is_expired(self) -> bool:
+        return datetime.now() > self.expires_at
```

Comment:
in the loop below and in the loop here i would put in some buffer that it rather requests it too early than too late. it can happen now that we get an error if we request something right when it expires if i understand this correctly? maybe have a 1 minute buffer?

Reply:
It refetches the token once the existing token is expired at 80%, this method is more used as a last resort

Reply:
okay, but that thing is still wrapped in a min(30)?  here:
                        sleep_time = min(30, int(self._current_token.time_until_refresh * (1 - self._refresh_threshold)))

or does it get decided elsewhere?

Reply:
I completely rewrote the logic